### PR TITLE
Avoid unnecessary coercion to Python on int comparisons

### PR DESCRIPTION
**Pull request by da-woods**
Fixes some of https://github.com/cython/cython/issues/4819 (although
there's a second issue about applying this more generally to binops).

It works by always generating a second utility code function with
a bint result type and substituting that out if needed. This is
on the basis that the coercion is likely to be sufficiently
common that the code generation is worth the effort

### CODE REVIEW

**Comment by scoder** [comment]
This implementation makes me wonder if replacing the `BinopNode` or `PrimaryCmpNode` with a `PythonCapiCallNode` is the right approach here. If we could leave the implementation to the nodes themselves, rather than the transform, it would be much more natural to select the right implementations in their `analyse_types()` and `coerce_to()` methods. And, honestly, the concrete implementation of the operator really belongs into the operator's tree node.

**Comment by da-woods** [comment]
> This implementation makes me wonder if replacing the `BinopNode` or `PrimaryCmpNode` with a `PythonCapiCallNode` is the right approach here. If we could leave the implementation to the nodes themselves, rather than the transform, it would be much more natural to select the right implementations in their `analyse_types()` and `coerce_to()` methods. And, honestly, the concrete implementation of the operator really belongs into the operator's tree node.

Possibly. I was trying to get the `cdef int x = 2+y` extension of this to work a few days ago, and it didn't work very easily under the scheme I'd got here. Moving it to the node like you're suggesting seems a lot more natural.

Shall I mark this as draft for now?

**Comment by scoder** [comment]
Yes, a draft seems right.

**Comment by da-woods** [comment]
I've moved some of the logic to `PrimaryCmpNode` but not done a big change. Essentially I've identified that the change can happen and set the return type to `bint` but still let `OptimizeBuiltinCalls` do the transform. Partly because there seems like a lot of code to move, and partly because the optimization is fairly specific - it's only about comparison with a constant rather than any operand with that type.

**Comment by da-woods** [comment]
Hmmm... this seems to have broken some stuff...

**Comment by da-woods** [comment]
Marking it as ready for review for now. There's a second linked change about trying to do a similar change for addition/multiplication, but not sure this needs to wait on it

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -13062,6 +13063,20 @@ def find_special_bool_compare_function(self, env, operand1, result_is_bool=False
...
+                        "Eq" if self.operator == "==" else "Ne",
+                        self,
+                        PyrexTypes.c_bint_type,
+                        self.operand1,
+                        self.operand2
+                    )
+                    if result:
+                        self.special_bool_cmp_function = result[0]
+                        self.special_bool_cmp_utility_code = result[1]
+                        self.special_bool_extra_args = result[2]
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:13078]: I find this slightly better.
```suggestion
                        (self.special_bool_cmp_function,
                         self.special_bool_cmp_utility_code,
                         self.special_bool_extra_args,
                         _) = result
```

File: Cython/Compiler/ExprNodes.py [hunk 2]
```diff
@@ -13111,14 +13126,19 @@ def generate_operation_code(self, code, result_code,
                 result2 = operand2.py_result()
             else:
                 result2 = operand2.result()
+            special_bool_extra_args_result = [
+                extra_arg.result() for extra_arg in self.special_bool_extra_args
+            ]
+            special_bool_extra_args_result = ", ".join(special_bool_extra_args_result)
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:13132]: Let's avoid using the same variable for two different things.
```suggestion
            special_bool_extra_args_result = ", ".join([
                extra_arg.result() for extra_arg in self.special_bool_extra_args
            ])
```

File: Cython/Compiler/Optimize.py [hunk 3]
```diff
@@ -4181,6 +4133,73 @@ def _inject_bint_default_argument(self, node, args, arg_index, default_value):
...
+            return None
+        numval = arg0
+        arg_order = 'CObj'
+    else:
+        return None
+
+    if not numval.has_constant_result():
+        return None
+
+    is_float = isinstance(numval, ExprNodes.FloatNode)
```
Comment (scoder) [inline Cython/Compiler/Optimize.py:4159]: This seems simpler (yes, I'm aware that this code was just moved around):
```suggestion
    is_float = numval.type.is_float
```

**Review by scoder — APPROVED** [review]
Nice, looks good.

**Pull request merged**